proxy/src/telemetry/metrics/prometheus.rs

```diff
@@ -0,0 +1,521 @@
+                    *metrics.response_duration(&labels) += end.since_response_open;
+                    *metrics.response_latency(&labels) += end.since_request_open;
+                });
+            },
+
+            Event::StreamResponseFail(ref res, ref fail) => {
+                // TODO: do we care about the failure's error code here, or
+                //       does res.status_code cover that?
+                let labels = self.root_labels
+                    .with_proxy_ctx(event.proxy())
+                    .with_response_ctx(res);
+                self.update(|metrics| {
+                    *metrics.response_total(&labels) += 1;
+                    *metrics.response_duration(&labels) += fail.since_response_open;
+                    *metrics.response_latency(&labels) += fail.since_request_open;
+                });
+            },
+
+            _ =>
+                // TODO: this is where we could count transport events,
+                //       if there were any metrics that cared about them.
+                // TODO: track request failures?
+                { },
+        };
+    }
+}
+
+
+// ===== impl Serve =====
+
+impl Serve {
+    fn new(metrics: &Arc<Mutex<Metrics>>) -> Self {
+        Serve { metrics: metrics.clone() }
+    }
+}
+
+impl HyperService for Serve {
+    type Request = HyperRequest;
+    type Response = HyperResponse;
+    type Error = hyper::Error;
+    type Future = FutureResult<Self::Response, Self::Error>;
+
+    fn call(&self, req: Self::Request) -> Self::Future {
+        if req.path() != "/metrics" {
+            return future::ok(HyperResponse::new()
+                .with_status(StatusCode::NotFound));
+        }
+
+        let body = {
+            let metrics = self.metrics.lock()
+                .expect("metrics lock poisoned");
+            format!("{}", *metrics)
+        };
+        future::ok(HyperResponse::new()
+            .with_header(ContentLength(body.len() as u64))
+            .with_header(ContentType::plaintext())
+            .with_body(body))
+    }
+}
+
+
+// ===== impl Labels =====
+
+impl From<Arc<ctx::Process>> for Labels {
+    fn from(_ctx: Arc<ctx::Process>) -> Self {
+        let inner = Arc::new(LabelsInner::new());
+        // TODO: when PR #448 lands, use CONDUIT_PROMETHEUS_LABELS to get the
+        //       owning pod spec labels.
+        Labels {
+            inner,
+        }
+    }
+}
+
+impl<'a> IntoIterator for &'a Labels {
+    type Item = (&'a str, &'a str);
+    // TODO: remove box.
+    type IntoIter = Box<Iterator<Item=Self::Item> + 'a>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.inner.into_iter()
+    }
+}
+
+impl Labels {
+
+    pub fn with_proxy_ctx<'a>(&self, proxy_ctx: &'a Arc<ctx::Proxy>) -> Labels {
+        let direction = if proxy_ctx.is_inbound() {
+            "inbound"
+        } else {
+            "outbound"
+        };
+        let inner = self.child() + ("direction", direction.to_owned());
```

Comment:
If the labels were a bunch of struct fields, then we wouldn't need to allocate and copy this string around a bunch of times, right? :D